app/gui/view/component-browser/component-group/src/multi.rs

```diff
@@ -0,0 +1,212 @@
+///     expression_accepted <+ group.expression_accepted.gate(&group.focused).map(move |e| (id, *e));
+/// }
+/// ```
+macro_rules! propagate_frp {
+    ($network:ident, $group:ident, $(($endpoint:ident, $expr:expr)),*) => {
+        frp::extend! { $network
+            $($endpoint <+ $group.$endpoint.gate(&$group.focused).map($expr);)*
+        }
+    }
+}
+
+
+
+// ===============
+// === Wrapper ===
+// ===============
+
+newtype_prim! {
+    /// An index of the component group.
+    GroupId(usize);
+}
+
+/// A wrapper around the FRP outputs of the several component groups.
+///
+/// Each field in this struct corresponds to certain component group FRP output with additional
+/// [`GroupId`] attached. [`GroupId`] is a simple index inside `groups` iterator passed to the
+/// [`Wrapper::new`] constructor. Only focused component groups propagate their events. At all
+/// times only a single group remains focused.
+///
+/// In addition to propagating FRP outputs, the [`Wrapper`] also emits `group.focus()` and
+/// `group.defocus()` when the mouse moves in or out the `group` shape.
+#[allow(missing_docs)]
+#[derive(Debug)]
+pub struct Wrapper {
+    pub selected_entry:            frp::Stream<Option<(GroupId, entry::Id)>>,
+    pub suggestion_accepted:       frp::Stream<(GroupId, entry::Id)>,
+    pub expression_accepted:       frp::Stream<(GroupId, entry::Id)>,
+    pub is_header_selected:        frp::Stream<(GroupId, bool)>,
+    pub header_accepted:           frp::Stream<GroupId>,
+    pub selection_position_target: frp::Stream<(GroupId, Vector2<f32>)>,
+    pub focused:                   frp::Stream<(GroupId, bool)>,
+}
+
+impl Wrapper {
+    /// Constructor.
+    pub fn new(scene: &Scene, network: &frp::Network, groups: impl Iterator<Item = Group>) -> Self {
+        let mouse = &scene.mouse.frp;
+        let groups = Rc::new(groups.collect::<Vec<_>>());
+        frp::extend! { network
+            mouse_in_group <- any(...);
+            selected_entry <- any(...);
+            suggestion_accepted <- any(...);
+            expression_accepted <- any(...);
+            is_header_selected <- any(...);
+            header_accepted <- any(...);
+            selection_position_target <- any(...);
+            focused <- any(...);
+        }
+
+        for (i, group) in groups.iter().enumerate() {
+            let id = GroupId::from(i);
+            frp::extend! { network
+                mouse_in <- map(&mouse.position,f!([scene,group](pos) {
+                    let pos_obj_space = group.to_object_space(&scene, *pos);
+                    group.is_inside(pos_obj_space)
+                }));
+                mouse_in_group <+ mouse_in.on_change().on_true().map(move |_| id);
+            }
+            match group {
+                Group::Regular(group) => {
+                    frp::extend! { network
+                        focused <+ group.focused.map(move |f| (id, *f));
```

Comment:
One worry here – searcher can contain dozens/hundreds of groups. We should rather not create all of them upfront, we might want to create/remove them while scrolling on-demand instead. Otherwise, we will have performance issues here. Is this taken into account?

Reply:
Good catch, I changed the design of the API slightly, now it has `add`/`remove` methods. The implementation was kinda reworked because of it.